src/dnx/dnx.cpp

```diff
@@ -226,150 +227,78 @@ bool ExpandCommandLineArguments(int nArgc, LPTSTR* ppszArgv, int& nExpandedArgc,
+
+    // Note: We use application base from DNX_APPBASE environment variable only if --appbase
+    // did not exist. if neither --appBase nor DNX_APPBASE existed we use current directory
+    if (!appBase)
     {
-        data.applicationBase = szCurrentDirectory;
+        appBase = GetAppBasePathFromEnvironment(buffer) ? buffer : currentDirectory.c_str();
     }
 
     // Prevent coreclr native bootstrapper from failing with relative appbase
-    TCHAR szFullAppBase[MAX_PATH];
-    if (!GetFullPath(data.applicationBase, szFullAppBase))
-    {
-        exitCode = 1;
-        goto Finished;
-    }
+    return GetFullPath(appBase, fullAppBasePath) != 0;
+}
 
-    data.applicationBase = szFullAppBase;
+int CallApplicationProcessMain(int argc, dnx::char_t* argv[], TraceWriter traceWriter)
+{
+    // Set the DNX_CONOSLE_HOST flag which will print exceptions to stderr instead of throwing
+    SetConsoleHost();
 
-    m_hHostModule = LoadNativeHost(pwzHostModuleName);
-    if (!m_hHostModule)
-    {
-        if (m_fVerboseTrace)
-            ::_tprintf_s(_T("Failed to load: %s\r\n"), pwzHostModuleName);
-        m_hHostModule = nullptr;
-        goto Finished;
-    }
-    if (m_fVerboseTrace)
-        ::_tprintf_s(_T("Loaded Module: %s\r\n"), pwzHostModuleName);
+    auto currentDirectory = GetNativeBootstrapperDirectory();
 
-    pfnCallApplicationMain = (FnCallApplicationMain)GetEntryPointFromHost(m_hHostModule, pszCallApplicationMainName);
-    if (!pfnCallApplicationMain)
-    {
-        if (m_fVerboseTrace)
-            ::_tprintf_s(_T("Failed to find function %S in %s\n"), pszCallApplicationMainName, pwzHostModuleName);
-        fSuccess = false;
-        goto Finished;
-    }
-    if (m_fVerboseTrace)
-        printf_s("Found DLL Export: %s\r\n", pszCallApplicationMainName);
+    // Set the DEFAULT_LIB environment variable to be the same directory as the exe
+    SetEnvironmentVariable(_T("DNX_DEFAULT_LIB"), currentDirectory.c_str());
 
-    hr = pfnCallApplicationMain(&data);
-    if (SUCCEEDED(hr))
-    {
-        fSuccess = true;
-        exitCode = data.exitcode;
-    }
-    else
-    {
-        fSuccess = false;
-        exitCode = hr;
-    }
+    CALL_APPLICATION_MAIN_DATA data = { 0 };
+    data.argc = argc;
+    data.argv = const_cast<const dnx::char_t**>(argv);
 
-Finished:
-    if (pfnCallApplicationMain)
+    dnx::char_t appBaseBuffer[MAX_PATH];
+    if (!GetApplicationBase(currentDirectory, argc, argv, appBaseBuffer))
     {
-        pfnCallApplicationMain = nullptr;
+        return 1;
     }
 
-    if (m_hHostModule)
-    {
-        if (FreeNativeHost(m_hHostModule))
-        {
-            fSuccess = true;
-        }
-        else
-        {
-            fSuccess = false;
-        }
-
-        m_hHostModule = nullptr;
-    }
+    data.applicationBase = appBaseBuffer;
 
-    if (bExpanded)
+    try
     {
-        FreeExpandedCommandLineArguments(nExpandedArgc, ppszExpandedArgv);
-    }
+        const dnx::char_t* hostModuleName =
```

Comment:
I did think about moving this elsewhere but I am on the edge here because it is being used only in this one place and currently we don't have a header for constants. I decided to leave it as is for now.